Add Acronym Support for Fuzzy Search

taooceros opened the pull request:
1. This pull request add acronym support for fuzzysearch. Link to #136 
2. This pr adds mapping from translated text to original text, allowing better highlighted text display for people using pinyin search (or potential future translation).

taooceros commented:
I originally want to do it as the way I have done before, but then I change my mind and decide to write a straight-forward search for the acronyms and query, which will make retrieve metadata easier and support the case capital letter and space both exist.

Because it is a straightforward loop, the performance should be good.

taooceros commented:
I think the test case need update

taooceros commented:
If this pull request has been approved, #183 should remove the part that create acronym before the pinyin. However, in order to achieve a correct matchdata highlight, IAlphabet should return an extra bool value suggesting whether the string is translated, since one Chinese character will project to multiple letter in pinyin form.

taooceros commented:
I am not sure how we should implement the data match for translated data, since the length change in translation...... should we simply don't highlight data when the string has been translated? Or return an extra string for matching data, which make it the same length as the original string.
Or create an extra class for mapping the index of translated string to original string, and create highlighted index based on that.

taooceros commented:
@jjw24 please take a try on the new mapping class, and it allow us to even highlight text when using full pinyin. However, it will take another loop to do the mapping every time, not sure if it will be possible to optimize it more.

jjw24 commented:
just need some clarification:
- is this acronym support for cases where 'vsc' gets 'visual studio code'?
- essentially it is for identifying first letter of every word?

taooceros commented:
> just need some clarification:
> - is this acronym support for cases where 'vsc' gets 'visual studio code'?
> - essentially it identifies first letter of every word?
> - should 'vs' for 'visual studio code' match?
> - currently this pr even match 'v' for 'visual studio code' which is a bit excessive

It matches the first letter of a word seperated from whitespace or upper letter. So vsc will match both "visual studio code" or "Visual Studio Code" or "visualStudioCode" or "VisualStudioCode".
Not sure whether we should make vs match for visual studio code. I think you are right that a small panalty should be added for unmatched after query finish, while a larger panalty has been added for unmatches before query end.

jjw24 commented:
When would there be a scenario where "visualStudioCode" or "VisualStudioCode" need to be acronym matched? program names are usually seperated by space, and if not the case it think should leave it to the fuzzy search to match up those kind of strings, that's what fuzzy matching is for.

I am concerned with performance when you have that loop at the beginning to do acronym search and you iterate through the whole 'stringtocompare' if that does not get a match, the rest of the code does another loop through the 'stringtocompare'.

i suggest we work it into the current loop. Take a look at the branch i pushed up:
https://github.com/Flow-Launcher/Flow.Launcher/compare/add_acronym_search?expand=1
https://github.com/Flow-Launcher/Flow.Launcher/tree/add_acronym_search

this has worked acronym search into the loop without another extra loop, it basically just projects forward to see the next substring if there is a match in case the loop is satisfied with the fuzzy search match and jumps to the next char.

I think it's a start, although it doesnt do the "visualStudioCode" or "VisualStudioCode" (which i think we should improve the actual fuzzy search code to do this) or where the the substring is numerical.

taooceros commented:
> When would there be a scenario where "visualStudioCode" or "VisualStudioCode" need to be acronym matched? program names are usually seperated by space, and if not the case it think should leave it to the fuzzy search to match up those kind of strings, that's what fuzzy matching is for.
> 
> I am concerned with performance when you have that loop at the beginning to do acronym search and you iterate through the whole 'stringtocompare' if that does not get a match, the rest of the code does another loop through the 'stringtocompare'.
> 
> i suggest we work it into the current loop. Take a look at the branch i pushed up:
> https://github.com/Flow-Launcher/Flow.Launcher/compare/add_acronym_search?expand=1
> (https://github.com/Flow-Launcher/Flow.Launcher/tree/add_acronym_search)
> 
> this has worked acronym search into the loop without another extra loop, it basically just projects forward to see if there is a match in the next substring in case the loop is satisfied with the fuzzy search match and jumps to the next char.
> 
> I think it's a start, although it doesnt do the "visualStudioCode" or "VisualStudioCode" (which i think we should improve the actual fuzzy search code to do this) or where the the substring is numerical.
> 
> Let me know what you think of this change

I added that because in my computer sometimes programs split there name like CamelCase or Pascal Case, for example "PostmanCanary" or "JetBrains". So I added that feature.

taooceros commented:
I will take a look on the branch, soon. However, I have encountered an issue that after merging dev branch, the setting windows is unable to get the dynamic resource, which most of the text in textblock disappears. I am not sure what cause that.....

jjw24 commented:
> I will take a look on the branch, soon. However, I have encountered an issue that after merging dev branch, the setting windows is unable to get the dynamic resource, which most of the text in textblock disappears. I am not sure what cause that.....

try delete output folder and rebuild

>I added that because in my computer sometimes programs split there name like CamelCase or Pascal Case, for example "PostmanCanary" or "JetBrains". So I added that feature.

hmm fair point, let me have a think how we can build it in.

taooceros commented:
> > I will take a look on the branch, soon. However, I have encountered an issue that after merging dev branch, the setting windows is unable to get the dynamic resource, which most of the text in textblock disappears. I am not sure what cause that.....
> 
> try delete output folder and rebuild

Let me take a try.

> 
> > I added that because in my computer sometimes programs split there name like CamelCase or Pascal Case, for example "PostmanCanary" or "JetBrains". So I added that feature.
> 
> hmm fair point, let me have a think how we can build it in.

I have checked the code for split right now, and it require an extra for loop to do that. So it will be the same for us to use our for loop to do the same jobs. However, I do think we can merge fuzzy search and acronym search since fuzzy search do perform a for loop. I don't do that because I haven't fully understand the full fuzzy search, so it make me a few scare to touch some important things. Let me take a try on that.

taooceros commented:
By the way, Marry Christmas! ❤

taooceros commented:
> try delete output folder and rebuild

That seems fix it. Thank you

jjw24 commented:
> By the way, Marry Christmas! ❤

Haha thank you. Merry Christmas to you too :)

jjw24 commented:
> I don't do that because I haven't fully understand the full fuzzy search, so it make me a few scare to touch some important things. Let me take a try on that.

No stress, we have a lot of tests around it.

>I have checked the code for split right now, and it require an extra for loop to do that.

Yeah I wonder if there is a way to do it without the split

jjw24 commented:
please resolve conflict @taooceros

taooceros commented:
Sorry I merge the wrong one, I will reset it soon.

taooceros commented:
ok, this should be great now...😅 @jjw24

jjw24 commented:
No problem will take it for a spin

jjw24 commented on Flow.Launcher.Infrastructure/StringMatcher.cs:
```diff
@@ -126,27 +177,39 @@ public MatchResult FuzzyMatch(string query, string stringToCompare, MatchOption
+                        AllQuerySubstringsMatched(currentQuerySubstringIndex, querySubstrings.Length);
+
                     if (allQuerySubstringsMatched)
-                        break;
+                        continue;
```
this change is correct?

taooceros replied:
Em yes, that is used to before early return of full match to check more acronym.

jjw24 commented on Flow.Launcher.Infrastructure/StringMatcher.cs:
```diff
@@ -43,21 +43,23 @@ public MatchResult FuzzyMatch(string query, string stringToCompare)
         /// </summary>
```
please update the FuzzyMatch summary to include the Acronym search steps

taooceros replied:
👌, and I have a question, should we add an option to open acronym search?

jjw24 replied:
sorry what do you mean open acronym search?

taooceros replied:
I mean we can have a parameters that indicates whether the search should include acronym search (maybe default is true).

jjw24 replied:
any reason why someone wouldnt use it or have the search method included?

taooceros replied:
Emmm I am not quite sure. For example, maybe we don't want acronym search used to matching for explorer plugin because the index search doesn't include acronym search.
Maybe for some description matching (generally in subtitle matching), we don't want acronym search.

jjw24 replied:
Good point, though Explorer's search can't utilize fuzzy search at all due to usage of DirectiryInfoSearch and Index, so no need to worry here. 

If we don't want description search might as well not use fuzzy on it either.

taooceros replied:
Well, I mean the Explorer utilize fuzzy search to create highlight data for results in result manager, so that may be a place that shouldn't use acronym search. ```

```

jjw24 replied:
but you can use acronym search with Explorer, so you cant type 'gk' and hit GitKraken, you have to type 'git' so highlight data would still be correct right?

taooceros replied:
Yeah I agree. I think there is some conner case but it is not a lot.

jjw24 commented on Flow.Launcher.Test/FuzzyMatcherTest.cs:
```diff
@@ -142,20 +144,20 @@ public void GivenQueryString_WhenAppliedPrecisionFiltering_ThenShouldReturnGreat
         [TestCase("goo", "Google Chrome", SearchPrecisionScore.Regular, true)]
         [TestCase("chr", "Google Chrome", SearchPrecisionScore.Low, true)]
         [TestCase("chr", "Chrome", SearchPrecisionScore.Regular, true)]
-        [TestCase("chr", "Help cure hope raise on mind entity Chrome", SearchPrecisionScore.Regular, false)]
```
why is this test case removed?

taooceros replied:
Because the acronym search will make this case matched, through `cure hope raise`. I think it is not the original idea for this test?

jjw24 replied:
This was to prevent fuzzy match bring up this kind of unrelated result. Maybe scoring needs adjusting so three acronyms shouldn't being up this long string?

taooceros replied:
emmm agree with you.

jjw24 commented on Flow.Launcher.Test/FuzzyMatcherTest.cs:
```diff
@@ -129,11 +131,11 @@ public void GivenQueryString_WhenAppliedPrecisionFiltering_ThenShouldReturnGreat
 
         [TestCase(Chrome, Chrome, 157)]
         [TestCase(Chrome, LastIsChrome, 147)]
-        [TestCase(Chrome, HelpCureHopeRaiseOnMindEntityChrome, 25)]
+        [TestCase(Chrome, HelpCureHopeRaiseOnMindEntityChrome, 90)]
```
this shouldnt have such a high score that it shows, acronym search matches it despite 'Help' at the start?

jjw24 replied:
hmm despite the penalty, i wonder if this is will bring up irrelevant results

taooceros replied:
I am not quite sure, but there shouldn't be few strings that will act like that. Maybe we can release a pre-release and check how people respond to it.

jjw24 commented on Flow.Launcher.Test/FuzzyMatcherTest.cs:
```diff
@@ -129,11 +131,11 @@ public void GivenQueryString_WhenAppliedPrecisionFiltering_ThenShouldReturnGreat
         [TestCase(Chrome, CandyCrushSagaFromKing, 0)]
-        [TestCase("sql", MicrosoftSqlServerManagementStudio, 110)]
-        [TestCase("sql  manag", MicrosoftSqlServerManagementStudio, 121)]//double spacing intended
+        [TestCase("sql", MicrosoftSqlServerManagementStudio, 90)]
```
this score has dipped due to acronym penalty?

taooceros replied:
yes, it match the middle "SQL" but without the M

jjw24 commented on Flow.Launcher.Test/FuzzyMatcherTest.cs:
```diff
@@ -142,20 +144,20 @@ public void GivenQueryString_WhenAppliedPrecisionFiltering_ThenShouldReturnGreat
-        [TestCase("ccs", "Candy Crush Saga from King", SearchPrecisionScore.Low, true)]
-        [TestCase("cand", "Candy Crush Saga from King",SearchPrecisionScore.Regular, true)]
-        [TestCase("cand", "Help cure hope raise on mind entity Chrome", SearchPrecisionScore.Regular, false)]
+        [TestCase("ccs", "Candy Crush Saga from King", SearchPrecisionScore.Regular, true)]
```
why has this precision changed to regular?

taooceros replied:
Because the acronym search will allow that to be matched with SearchPrecisionScore.Regular

jjw24 commented on Flow.Launcher.Test/FuzzyMatcherTest.cs:
```diff
@@ -196,31 +201,38 @@ public void GivenQueryString_WhenAppliedPrecisionFiltering_ThenShouldReturnGreat
         [TestCase("sql studio", MicrosoftSqlServerManagementStudio, SearchPrecisionScore.Regular, true)]
         [TestCase("mic", MicrosoftSqlServerManagementStudio, SearchPrecisionScore.Regular, true)]
+        [TestCase("mssms", MicrosoftSqlServerManagementStudio, SearchPrecisionScore.Regular, true)]
+        [TestCase("msms", MicrosoftSqlServerManagementStudio, SearchPrecisionScore.Regular, true)]
```
hmm ok, so acronym match doesnt have to match entire string

taooceros replied:
Well, yes, for now. Because we include description in searching for now, which means that there will be quite a lot of unmatched acronym if we match the entire string.

jjw24 commented on Flow.Launcher/ViewModel/MainViewModel.cs:
```diff
@@ -136,6 +121,22 @@ void continueAction(Task t)
             }
         }
 
+        private void RegisterResultsUpdatedEvent()
+        {
+            foreach (var pair in PluginManager.GetPluginsForInterface<IResultUpdated>())
+            {
+                var plugin = (IResultUpdated)pair.Plugin;
+                plugin.ResultsUpdated += (s, e) =>
+                {
+                    if (e.Query.RawQuery == QueryText) // TODO: allow cancellation
+                    {
+                        PluginManager.UpdatePluginMetadata(e.Results, pair.Metadata, e.Query);
+                        _resultsUpdateQueue.Post(new ResultsForUpdate(e.Results, pair.Metadata, e.Query, _updateToken));
+                    }
+                };
+            }
+        }
+
```
what's this change?

taooceros replied:
That's due to unintended merge. It seems just swap the posible of the two registration.

jjw24 replied:
There is a bit of difference, can you please double check including it is ok

jjw24 reviewed (changes requested):
please update summary

jjw24 commented:
Feels like we should adjust the scoring. With the current code, is it possible to score based on percentage of acronyms matched? So three characters is not going to bring up an entire long phrase

taooceros commented:
> Feels like we should adjust the scoring. With the current code, is it possible to score based on percentage of acronyms matched? So three characters is not going to bring up an entire long phrase

Yes, I agree with you. However, I think we still match the description in Program plugin, which means some description is long? Maybe we shall seperate them, with matching title with acronym and description without acronym, and add panelty for mismatch after the query is finished for acronym.

taooceros commented:
> please update summary

Will do tmr, or on the weekend.

jjw24 commented:
> > Feels like we should adjust the scoring. With the current code, is it possible to score based on percentage of acronyms matched? So three characters is not going to bring up an entire long phrase
> 
> Yes, I agree with you. However, I think we still match the description in Program plugin, which means some description is long? Maybe we shall seperate them, with matching title with acronym and description without acronym, and add panelty for mismatch after the query is finished for acronym.
> 
> # e.g.
> ![image](https://user-images.githubusercontent.com/45326534/106138643-3e2ec680-6121-11eb-95c9-487e06d85aa7.png)

i dont think the problem is whether we match description or not, in general acronym shouldnt bring up irrelevant long strings. That's why i was thinking maybe scoring needs to be percentage based. I dont think we should specifically exclude acronym from description because this will start to get confusing, not to mention the string matcher class is getting dangerously unmaintainable. 

What you have with penalty may work as well, and for long strings penalty could be so large that the acronym score will go into negative, which is fine as users can still use rest of fuzzy search method to get the results if they were to look for a long string.

jjw24 commented:
Just reworking scoring to be based on percentage not total acronyms so code doesn't bring up long strings. Will push up the branch for you to take a look

taooceros commented:
Thank you!

jjw24 commented on Flow.Launcher.Infrastructure/StringMatcher.cs:
```diff
@@ -72,19 +87,63 @@ public MatchResult FuzzyMatch(string query, string stringToCompare, MatchOption
+                if (char.IsUpper(stringToCompare[compareStringIndex]) ||
+                    char.IsNumber(stringToCompare[compareStringIndex]) ||
+                    char.IsWhiteSpace(stringToCompare[compareStringIndex]) ||
+                    spaceMet)
```
@taooceros what's this 'spaceMet' variable trying to solve and is this needed for translation?

taooceros replied:
I think it is used to mark character after space to be acronym even when it is lower case.

jjw24 replied:
Ok cool, I thought so as well. I am thinking of using substring index = 0 to replace this, since the start of substring char is always acronym

taooceros replied:
Have we split the string to compare? I think we only split the query string.

jjw24 replied:
Oh we didn't, lol ok nvm then

jjw24 replied:
Alternatively then, we can go back an index to check if it is space

taooceros replied:
Yeah, that's feasible.

jjw24 commented:
please look at the updates and let me know if good to merge in here:
https://github.com/taooceros/Flow.Launcher/pull/4/files

taooceros commented:
> please look at the updates and let me know if good to merge in here:
> https://github.com/taooceros/Flow.Launcher/pull/4/files

Great! I will take a look soon.

taooceros commented:
@jjw24 any further change we need to make for this PR?

jjw24 commented:
Let me take a final look in 2hrs

jjw24 reviewed (approved):
Great work @taooceros 👍
